Ctrl-A only increments number in path segment (Fixes #850)

Kingdread opened the pull request:
This prevents a host like "myfoo42.bar" changing to "myfoo43.bar" when
pressing Ctrl-A. It further prevents increasing the port number, e.g.
going from "foo.bar:8080" to "foo.bar:8081".

This should fix #850.

The-Compiler commented:
Thank you, and congrats on your first contribution! :smile:

Now that I look at it, I think the whole logic should be moved to `utils/urlutils.py` and tests should be added to `tests/utils/test_urlutils.py`. Do you want to do that or should I? I'm of course happy to help if you have issues!

Kingdread commented:
Thanks!

I guess I could do that, if I have any issues I will come back to you and ask for your help :D

The-Compiler commented:
Awesome!

Some remarks:

- I have no idea why I did it via `toEncoded()` before - this means something like http://www.qutebrowser.org/utf8=%E2%9C%93 will increment the character - so that'd make a good test! :laughing: 
- Basically everything of `_navigate_incdec` can go to urlutils, except the last line of course.
- You should define a new exception which makes sense, say `IncDecError` or whatever and use that where `CommandError` is used - and then in the command do something like:

```
try:
    urlutils.inc_dec(url)   # or whatever
except urlutils.IncDecError as e:
    raise cmdexc.CommandError(e)
```

Kingdread commented:
I called the function `url_incdec_number` and added some tests to `tests/utils/test_urlutils.py`. I've also added a case for the urlencoded data (`%C3%A4`) :smile:

The-Compiler commented on qutebrowser/utils/urlutils.py:
```diff
@@ -443,3 +443,60 @@ def __str__(self):
+        return '{}: {}'.format(self.msg, self.url.toString())
+
+
+def url_incdec_number(url, incdec):
```
What do you think about removing the `url_` prefix? It kinda seems redundant to me as it's in `urlutils`.

The-Compiler commented on qutebrowser/utils/urlutils.py:
```diff
@@ -443,3 +443,60 @@ def __str__(self):
+        The new url with the number incremented/decremented.
+
+    Raises IncDecError if the url contains no number.
+    """
```
Please add something like this (and a test for it):

```python
    if not url.isValid():
        raise ValueError(get_errstring(url))
```

This isn't strictly necessary (as it's already enforced by `_current_url()`), but since it's now part of `urlutils` and might be used in other places, let's add it.

The-Compiler commented on qutebrowser/utils/urlutils.py:
```diff
@@ -443,3 +443,60 @@ def __str__(self):
+        raise IncDecError("No number found in URL!", url)
+    pre, number, post = match.groups()
+    if not number:
+        raise IncDecError("No number found in URL!", url)
```
Hm. These two lines aren't tested (and `urlutils.py` is one of the modules where I enforce 100% coverage), but I really can't see why. `(\d+)` can't possibly be an empty string, right?

If you agree, let's just remove them.

The-Compiler commented on qutebrowser/utils/urlutils.py:
```diff
@@ -443,3 +443,60 @@ def __str__(self):
+    try:
+        val = int(number)
+    except ValueError:
+        raise IncDecError("Could not parse number '{}'.".format(number), url)
```
Same thing here essentially - I don't think this can ever happen with a string matching `\d+`.

The-Compiler commented on qutebrowser/utils/urlutils.py:
```diff
@@ -443,3 +443,60 @@ def __str__(self):
+    elif incdec == 'increment':
+        val += 1
+    else:
+        raise ValueError("Invalid value {} for indec!".format(incdec))
```
Can you add a test for this please?

The-Compiler commented on qutebrowser/utils/urlutils.py:
```diff
@@ -443,3 +443,60 @@ def __str__(self):
+        self.msg = msg
+
+    def __str__(self):
+        return '{}: {}'.format(self.msg, self.url.toString())
```
Can you add a test for this please?

The-Compiler commented:
Thanks again! I added some more comments.

Can you please also add an entry to `CHANGELOG.asciidoc`? Something like *:navigate with increment/decrement argument (`<Ctrl-A>`/`<Ctrl-X>`) now handles some corner-cases better.* or so under *Fixed:* should do.

Kingdread commented:
Updated!

I agree with the regex change, `(\d+)` should never return an empty string and the matched part should always be convertible to an `int` - even if it contains some weird Unicode digits (`int("\u0A68")` returns 2, the char is "GURMUKHI DIGIT TWO").

The-Compiler commented on tests/utils/test_urlutils.py:
```diff
@@ -524,3 +524,63 @@ def test_same_domain_invalid_url(url1, url2):
     with pytest.raises(ValueError):
         urlutils.same_domain(QUrl(url1), QUrl(url2))
+
+@pytest.mark.parametrize('url, incdec, output', [
```
I think it'd be better if there was a `TestIncdec` class to group those, and then they can be methods without the `incdec_` prefix. I usually do this if there are many tests which belong together.

(sidenote: I know I'm a perfectionist - if you get too annoyed by this kind of thing, please tell me! Then I'll do such style-changes myself instead. But I think this is the last one other than the test mentioned in #850 :wink:)

Kingdread replied:
I'm fine with perfectionism, feel free to point out the stuff I need to change :smile:

The-Compiler commented:
Merged - thank you very much!
